linera-core/src/chain_worker/cache.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Types to handle a set of running chain workers.
+//!
+//! The [`ChainWorkers`] type contains a set of running [`ChainWorkerActor`]s. It will
+//! limit itself to a maximum number of active tasks, and will evict the least-recently
+//! used chain worker when it is full and a new chain worker actor needs to be created.
+
+use std::{
+    collections::{HashMap, HashSet},
+    num::NonZeroUsize,
+    sync::{Arc, Mutex, RwLock},
+};
+
+use linera_base::{crypto::CryptoHash, identifiers::ChainId};
+use linera_chain::{data_types::ExecutedBlock, types::Hashed};
+use linera_storage::Storage;
+use lru::LruCache;
+use tokio::sync::{mpsc, OwnedSemaphorePermit, Semaphore};
+use tracing::Instrument as _;
+
+use crate::{
+    chain_worker::{ChainWorkerActor, ChainWorkerConfig, ChainWorkerRequest, DeliveryNotifier},
+    join_set_ext::{JoinSet, JoinSetExt},
+    value_cache::ValueCache,
+    worker::WorkerError,
+};
+
+/// A cache of running [`ChainWorkerActor`]s.
+#[derive(Clone)]
+pub struct ChainWorkers<StorageClient>
+where
+    StorageClient: Storage,
+{
+    /// One-shot channels to notify callers when messages of a particular chain have been
+    /// delivered.
+    delivery_notifiers: Arc<Mutex<DeliveryNotifiers>>,
+    /// The set of spawned [`ChainWorkerActor`] tasks.
+    tasks: Arc<Mutex<JoinSet>>,
+    /// The cache of running [`ChainWorkerActor`]s.
+    cache: Arc<Mutex<LruCache<ChainId, ChainRequestSender<StorageClient>>>>,
+    /// Active endpoint permits.
+    active_endpoints: Arc<Semaphore>,
+}
+
+/// The map of [`DeliveryNotifier`]s for each chain.
+pub(crate) type DeliveryNotifiers = HashMap<ChainId, DeliveryNotifier>;
+
+impl<StorageClient> ChainWorkers<StorageClient>
+where
+    StorageClient: Storage,
+{
+    /// Creates a new [`ChainWorkers`] that stores at most `limit` workers.
+    pub fn new(limit: NonZeroUsize) -> Self {
+        ChainWorkers {
+            delivery_notifiers: Arc::default(),
+            tasks: Arc::default(),
+            cache: Arc::new(Mutex::new(LruCache::new(limit))),
+            active_endpoints: Arc::new(Semaphore::new(limit.get())),
```

Comment:
Does using the same limit here mean that if `limit` requests are in flight to a single chain worker, and there's another incoming request for a different worker, we will now block that new request even though a slot would be free?

Reply:
Good catch! It does seem like we're first trying to acquire a `self.active_endpoints` Semaphore even if there already exists a chain worker for the same `chain_id`.

Reply:
Yes, that is true. I thought about having a map of permits, so that requests that get queued anyway would reuse the existing permit. However, that may lead to starvation because a chain worker that is heavily used will never release its permit.

The solution I thought then was to have another semaphore for acquiring each permit, but I thought the permitception would be overengineering. I'm not sure how to proceed here, and would love some feedback and ideas.

I still think it's okay to improve (fix?) this in a follow-up PR because:
- this impacts a (hopefully) edge case (cache is full)
- worst outcome is degraded performance
- with the benefit that it would be fairer in extreme scenarios

Reply:
I don't think it's an edge case to have several requests for the same chain worker. And every time we have that we underutilize at least one other chain worker.

But I don't have a better idea at the moment either.